File: Ex23_Minimum_Spanning_Tree/codes/boruvka.py
if __name__ == "__main__":
    n, m = map(int, input().split())
    edges = []
    for _ in range(m):
        u, v, w = map(int, input().split())
        edges.append((w, u, v))

    parent = [i for i in range(n + 1)]
    rank = [0 for _ in range(n + 1)]

    def find(u):
        while u != parent[u]:
            u = parent[u]
        return u

    def union(u, v):
        ru, rv = find(u), find(v)
        if ru == rv:
            return False
        if rank[ru] < rank[rv]:
            parent[ru] = rv
        else:
            parent[rv] = ru
            if rank[ru] == rank[rv]:
                rank[ru] += 1
        return True

    total = 0
    num_comp = n

    while num_comp > 1:
        cheapest = [None for _ in range(n + 1)]

        for w, u, v in edges:
            ru, rv = find(u), find(v)
            if ru == rv:
                continue
            if cheapest[ru] is None or cheapest[ru][0] > w:
                cheapest[ru] = (w, u, v)
            if cheapest[rv] is None or cheapest[rv][0] > w:
                cheapest[rv] = (w, u, v)

        for i in range(1, n + 1):
            if cheapest[i] is not None:
                w, u, v = cheapest[i]
                if union(u, v):
                    total += w
                    num_comp -= 1

    print(total)

File: Ex23_Minimum_Spanning_Tree/codes/boruvka.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>

using namespace std;

const int MAXN = 100005;

int parent[MAXN], rank_arr[MAXN];
int n, m;

int find(int u) {
    while (u != parent[u]) {
        u = parent[u];
    }
    return u;
}

bool union_set(int u, int v) {
    int ru = find(u);
    int rv = find(v);
    if (ru == rv) return false;
    if (rank_arr[ru] < rank_arr[rv]) {
        parent[ru] = rv;
    } else {
        parent[rv] = ru;
        if (rank_arr[ru] == rank_arr[rv]) rank_arr[ru]++;
    }
    return true;
}

int main() {
    cin >> n >> m;
    vector<tuple<int, int, int>> edges(m);
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        edges[i] = {w, u, v};
    }

    for (int i = 1; i <= n; ++i) {
        parent[i] = i;
        rank_arr[i] = 0;
    }

    int total = 0;
    int num_comp = n;

    while (num_comp > 1) {
        vector<tuple<int, int, int>*> cheapest(n + 1, nullptr);

        for (auto& edge : edges) {
            int w = get<0>(edge), u = get<1>(edge), v = get<2>(edge);
            int ru = find(u);
            int rv = find(v);
            if (ru == rv) continue;

            if (!cheapest[ru] || get<0>(*cheapest[ru]) > w)
                cheapest[ru] = &edge;
            if (!cheapest[rv] || get<0>(*cheapest[rv]) > w)
                cheapest[rv] = &edge;
        }

        for (int i = 1; i <= n; ++i) {
            if (cheapest[i]) {
                int w = get<0>(*cheapest[i]);
                int u = get<1>(*cheapest[i]);
                int v = get<2>(*cheapest[i]);
                if (union_set(u, v)) {
                    total += w;
                    num_comp--;
                }
            }
        }
    }

    cout << total;
    return 0;
}
